### PR TITLE
KAFKA-14908: Set setReuseAddress on the kafka server socket

### PR DESCRIPTION
**Pull request by k-wall**
Changes `SocketServer` to set the `setReuseAddress(true)` socket option.

This aids use-cases where kafka is started/stopped on the same port in rapid succession. Examples are: where a kafka cluster is embedded in an integration test suite that starts/stops a cluster before/after each test.

I don't think this change brings any backward compatibility concerns, so I don't think it will need to be configurable.  However I can make it that way if necessary.

I have tested this change locally and confirmed that with the option the sporadic issue I see goes away.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by tombentley — APPROVED** [review]

**Review by divijvaidya — CHANGES_REQUESTED** [review]
Thank you for the change. 

First, I will rephrase the motivation so that we are on the same page:

`SO_REUSEADDR` helps re-use socket when it is in `TIME_WAIT` state. A socket reaches `TIME_WAIT` state to ensure clean disconnect of TCP connections. This `TIME_WAIT` will impact scenarios where: 1\ client is creating large number of short lived connections 2\ server  (or another process) is rapidly connecting & disconnecting from the same socket.

It's rare to see this scenario#1 in production Kafka because users don't typically create thousands of short lived connections. Kafka connections are usually long running due to cost involved in setting up a connection. We are not making a change in this PR to client side socket configuration Or the accepted connection socket configuration (at `SocketServer#configureAcceptedSocketChannel`), hence, scenario#1 is not fixed by this PR.

It's rare to see this scenario#2 in production Kafka because Kafka server clean shutdown process and the startup is quite long (since we perform recovery etc. on startup). Hence, this fix will impact cases such as integration tests you mentioned where we are rapidly shutting down and starting up server.

Next, I would like to discuss a couple of risks that this change may introduce so that we have considered the impact of this change for all users.

1. A socket sitting in TIME_WAIT state is intentional. It exists to ensure clean disconnect of TCP connections. We are ignoring this clean disconnection when we set `SO_REUSEADDR`. It can cause unintended side effects since the socket may technically still be considered to be in-use by the remote client. We might end up receiving messages which were intended for the previous connection. This risk seems acceptable to me since it's the same client (same IP/same port) who will be making the new connection. Also, we are only configuring this for listening socket and not for accepted sockets.

2. Security: SO_REUSEADDR can allow an attacker to perform DDOS by creating a genuine connection with same IP and port. Prior to this change, the TIME_WAIT state of the socket would have prevented immediate re-connection. This is an acceptable risk for Kafka because we have connection throttling available in code for IP addresses and a user may choose to configure it to prevent a DDOS.

The risks above are acceptable and other Apache projects such as [thrift server](https://github.com/apache/thrift/blob/1e3d90d8fd4160d538b7a4d902169eae5155e08a/lib/java/src/main/java/org/apache/thrift/transport/TServerSocket.java#L89), [tomcat](https://github.com/apache/tomcat/blob/5afdd20934596e5d19e45364e8e883b61f9d1daa/java/org/apache/catalina/ha/backend/TcpSender.java#L101) etc. are already using this configuration. I think this is a useful change for Kafka server and clients.

Before we approve this PR, I would request you to change the following in the code:
1. Update the tests with this configuration where ever we are opening a SocketServerChannel such as SelectorTest.java, SSLSelectorTest.java etc.
2. I would suggest to extract out a method "createSocketServerChannel()" in SocketServer.java. This method would create the channel and set the configuration on the socket. Use this same method in all the tests. 
3. Add a comment in the code explaining the reason why are we setting this configuration.
4. Add a test which starts up the server and validates that the socket bound to listening port has `SO_REUSEADDR` enabled. This test will fail before this change and success afterwards.

**Comment by k-wall** [comment]
@divijvaidya thanks for taking time to put together the detailed reply, I'll make the changes you are suggesting.

**Comment by showuon** [comment]
@divijvaidya , thanks for the comment. I agree with you most of them. But for this point:
> I would suggest to extract out a method "createSocketServerChannel()" in SocketServer.java. This method would create the channel and set the configuration on the socket. Use this same method in all the tests.

Since the `SelectorTest` and other network tests are in client module, we should not add `core` dependency to `client` module. So, I think adding a util method in `TestUtils` should be good enough. WDYT?

**Comment by k-wall** [comment]
> @divijvaidya , thanks for the comment. I agree with you most of them. But for this point:
> 
> > I would suggest to extract out a method "createSocketServerChannel()" in SocketServer.java. This method would create the channel and set the configuration on the socket. Use this same method in all the tests.
> 
> Since the `SelectorTest` and other network tests are in client module, we should not add `core` dependency to `client` module. So, I think adding a util method in `TestUtils` should be good enough. WDYT?

`NioEchoServer`,  `SelectorTest `and `SslSelectorTests`  already make sensible use of `bind(0)`, so the listening sockets they create cannot actually suffer the problem described in the JIRA, even if those tests were run in tight loop.  The network stack  would chose an ephemeral port that was not in timed wait.    I think there is an argument for leaving these tests untouched.

I'm +1 to adding a new core test to ensure that the listener sockets made by Acceptors  do have SO_REUSEADDR set correctly and of course the comment to explain why it is so.

**Comment by divijvaidya** [comment]
@showuon That's a good point. The problem I want to solve here is to ensure that tests are using the same code that SocketServer is using. If we implement a method in `TestUtils` then the tests may not catch bugs when `SocketServer` was modified but `TestUtils` wasn't. But I agree with your concern about taking a dependency on core.

I agree with ignoring the part of my suggestion to modify tests (suggestion 2 in my previous comment) for now. But let's keep the part about adding a unit test to verify that this configuration was added (suggestion 4 in my previous comment).

File: core/src/test/scala/unit/kafka/network/SocketServerTest.scala
Comment (divijvaidya) [inline core/src/test/scala/unit/kafka/network/SocketServerTest.scala]: perhaps do this in try/finally so that the resources are cleaned even if the test fails with an exception.

File: core/src/test/scala/unit/kafka/network/SocketServerTest.scala
Comment (divijvaidya) [inline core/src/test/scala/unit/kafka/network/SocketServerTest.scala:1909]: please add a comment here.

We are using 0 so that OS can choose to associated any available port here.

File: core/src/test/scala/unit/kafka/network/SocketServerTest.scala
Comment (divijvaidya) [inline core/src/test/scala/unit/kafka/network/SocketServerTest.scala]: since we are starting up two listeners, could we verify reuse address for both please.

**Review by divijvaidya — CHANGES_REQUESTED** [review]
Thank you for the making the suggested changes! 

I added some last few comments to make the test more resilient in cleaning up resources.

File: core/src/test/scala/unit/kafka/network/SocketServerTest.scala
Comment (k-wall) [reply core/src/test/scala/unit/kafka/network/SocketServerTest.scala]: Now using try/finally.

File: core/src/test/scala/unit/kafka/network/SocketServerTest.scala
Comment (k-wall) [reply core/src/test/scala/unit/kafka/network/SocketServerTest.scala]: Test renamed and test now checks both.

File: core/src/test/scala/unit/kafka/network/SocketServerTest.scala
Comment (k-wall) [reply core/src/test/scala/unit/kafka/network/SocketServerTest.scala:1909]: Comment added documenting use of ephemerals.

**Review by divijvaidya — APPROVED** [review]
Thank you again for the quick turn around. Looks good to me. 
I also noticed that this is your first change to Apache Kafka. Congratulations! 🎉

Next step, a committer will look into this PR and merge it. I am guessing @tombentley or @showuon will be interested here.

**Comment by k-wall** [comment]
@divijvaidya thanks for the review feedback, much appreciated.

**Review by showuon — APPROVED** [review]
LGTM! Thanks for the patch.

**Comment by cmccabe** [comment]
Hi all,

Unfortunately I had to revert this. This is a change to our public API since it will allow multiple Kafka servers to be started on the same port. This kind of change needs a KIP (Kafka Improvement Proposal) so we can discuss the pros and cons.

The change also does very different things on different operating systems. There is a rundown here: https://stackoverflow.com/questions/14388706/how-do-so-reuseaddr-and-so-reuseport-differ

> Security: SO_REUSEADDR can allow an attacker to perform DDOS by creating a genuine connection with same IP and port. Prior to this change, the TIME_WAIT state of the socket would have prevented immediate re-connection. This is an acceptable risk for Kafka because we have connection throttling available in code for IP addresses and a user may choose to configure it to prevent a DDOS.

Again, this needs a KIP so we can discuss whether the increased DDOS risk is acceptable or not. The answer may be different for different users.

Personally, I think that at minimum SO_REUSEADDR should be an option (not mandatory) and probably not the default.

**Comment by divijvaidya** [comment]
Hey @cmccabe, we can have a community discussion over the mailing list but I am not following one of your statements here.

> since it will allow multiple Kafka servers to be started on the same port

Can you please help me understand why? I am asking this because AFAIK (correct me if I am wrong), all that this flag does is to allow re-use of a socket which is in TIME_WAIT state. If it is in another state, i.e. it is being used by a different Kafka process, the binding is not allowed, hence, two Kafka servers cannot run on the same port (or rather in this context, reuse the socket).

**Comment by k-wall** [comment]
Hello @cmccabe, thank you for the feedback.

> since it will allow multiple Kafka servers to be started on the same port

Having read [that SO article](https://stackoverflow.com/questions/14388706/how-do-so-reuseaddr-and-so-reuseport-differ) I now see the SO_REUSEADDR does indeed have that effect on some platforms.  It was not my intent to allow that. I was unknowingly fixating on the Linux behaviour.

For example, SO article discusses BSD behaviour:

> Without SO_REUSEADDR, binding socketA to 0.0.0.0:21 and then binding socketB to 192.168.0.1:21 will fail (with error EADDRINUSE), since 0.0.0.0 means "any local IP address", thus all local IP addresses are considered in use by this socket and this includes 192.168.0.1, too. With SO_REUSEADDR it will succeed, since 0.0.0.0 and 192.168.0.1 are not exactly the same address, one is a wildcard for all local addresses and the other one is a very specific local address. Note that the statement above is true regardless in which order socketA and socketB are bound; without SO_REUSEADDR it will always fail, with SO_REUSEADDR it will always succeed.

> Personally, I think that at minimum SO_REUSEADDR should be an option (not mandatory) and probably not the default.

I'd be happy to resubmit the PR in this way.

**Pull request merged**
